frontend/src/metabase/nav/hooks.unit.spec.tsx

```diff
@@ -0,0 +1,195 @@
+import fetchMock from "fetch-mock";
+
+import { renderWithProviders, screen, waitFor } from "__support__/ui";
+import { setupEnterpriseTest } from "__support__/enterprise";
+import { mockSettings } from "__support__/settings";
+import { setupDatabasesEndpoints } from "__support__/server-mocks";
+import {
+  createMockDatabase,
+  createMockTokenFeatures,
+  createMockUser,
+} from "metabase-types/api/mocks";
+import { createMockState } from "metabase-types/store/mocks";
+import { createSampleDatabase } from "metabase-types/api/mocks/presets";
+
+import { useShouldShowDatabasePromptBanner } from "./hooks";
+
+interface setupOpts {
+  isAdmin?: boolean;
+  isPaidPlan?: boolean;
+  onlyHaveSampleDatabase?: boolean;
+  isOnAdminAddDatabasePage?: boolean;
+  isWhiteLabeling?: boolean;
+}
+const TEST_DB = createSampleDatabase();
+
+const DATA_WAREHOUSE_DB = createMockDatabase({ id: 2 });
+
+async function setup({
+  isAdmin = false,
+  isPaidPlan = false,
+  isWhiteLabeling = false,
+  onlyHaveSampleDatabase = false,
+}: setupOpts = {}) {
+  if (onlyHaveSampleDatabase) {
+    setupDatabasesEndpoints([TEST_DB]);
+  } else {
+    setupDatabasesEndpoints([TEST_DB, DATA_WAREHOUSE_DB]);
+  }
+
+  const state = createMockState({
+    currentUser: createMockUser({ is_superuser: isAdmin }),
+    settings: mockSettings({
+      "token-features": createMockTokenFeatures(
+        isPaidPlan ? { sso: true } : {},
+      ),
+      "application-name": isWhiteLabeling ? "Acme Corp." : "Metabase",
+    }),
+  });
+
+  function TestComponent() {
+    const shouldShowDatabasePromptBanner = useShouldShowDatabasePromptBanner();
+    return shouldShowDatabasePromptBanner ? (
+      <>showing database prompt banner</>
+    ) : (
+      <>hiding database prompt banner</>
+    );
+  }
+
+  renderWithProviders(<TestComponent />, {
+    storeInitialState: state,
+  });
+
+  // 1. We will only call this endpoint when `isAdmin` and `isPaidPlan` are both true.
+  // 2. This check ensures the conditions for database prompt banner are all available.
+  // Then we could safely assert that the banner is not rendered.
+  // If we don't wait for this API call to finish, the banner could have rendered,
+  // and the test would still pass.
+  if (isAdmin && isPaidPlan && !isWhiteLabeling) {
+    await waitFor(() => {
+      expect(fetchMock.called("path:/api/database")).toBe(true);
+    });
+  }
```

Comment:
I think that lots of confusion comes from the fact there always is a default value for `databases = []` in `useShouldShowDatabasePromptBanner`. Let's try to think what would happen in `useShouldShowDatabasePromptBanner` without the default value.

Until the API call is triggered and completed `databases` would have been `undefined`.
This way `onlyHaveSampleDatabase` would also need to have been `undefined` and therefore the return value of this hook would also have been `undefined` (note: this is actually semantically correct).
This way we would know that we don't know whether to render the banner - we could render a Loader in the `TestComponent` and wait for it to disappear.

I feel like it would simplify things.

Reply:
I really hate boolean with 3 states (`undefined`/ `true`/ `false`), but this could solve the problem we have since the 2 states (`true` and `false`) won't tell whether or not the API is completed.

Good advise. I think I could try to make that work.